# Assignment 5: Brandi Beals
## MSDS 422 DL-55
You will compete in the Kaggle.com Digit Recognizer competition which involves classical digit recognition from hand-written images. https://www.kaggle.com/c/digit-recognizer/ 

Read the competition rules, and download the MNIST training and test set data.This binary classification task is NOT what is required for the current assignment. In this assignment we are asking for a multiclass classifier. The entire MNIST data set will be used for input data. For this assignment, you will develop a classifier that may be used to predict which of the 10 digits is being written.

In [21]:
# import libraries
import numpy as np
import pandas as pd
import datetime

In [2]:
# read in dataset
data = pd.read_csv('digit-recognizer/train.csv')
test_data = pd.read_csv('digit-recognizer/test.csv')

In [3]:
data.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42000 entries, 0 to 41999
Columns: 785 entries, label to pixel783
dtypes: int64(785)
memory usage: 251.5 MB


In [5]:
data.describe()

label   pixel0   pixel1   pixel2   pixel3   pixel4   pixel5  \
count  42000.000000  42000.0  42000.0  42000.0  42000.0  42000.0  42000.0   
mean       4.456643      0.0      0.0      0.0      0.0      0.0      0.0   
std        2.887730      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.000000      0.0      0.0      0.0      0.0      0.0      0.0   
25%        2.000000      0.0      0.0      0.0      0.0      0.0      0.0   
50%        4.000000      0.0      0.0      0.0      0.0      0.0      0.0   
75%        7.000000      0.0      0.0      0.0      0.0      0.0      0.0   
max        9.000000      0.0      0.0      0.0      0.0      0.0      0.0   

        pixel6   pixel7   pixel8  ...      pixel774      pixel775  \
count  42000.0  42000.0  42000.0  ...  42000.000000  42000.000000   
mean       0.0      0.0      0.0  ...      0.219286      0.117095   
std        0.0      0.0      0.0  ...      6.312890      4.633819   
min        0.0      0.0      0.0  ...      0.000000      0.000000   
25%        0.0      0.0      0.0  ...      0.000000      0.000000   
50%        0.0      0.0      0.0  ...      0.000000      0.000000   
75%        0.0      0.0      0.0  ...      0.000000      0.000000   
max        0.0      0.0      0.0  ...    254.000000    254.000000   

           pixel776     pixel777      pixel778      pixel779  pixel780  \
count  42000.000000  42000.00000  42000.000000  42000.000000   42000.0   
mean       0.059024      0.02019      0.017238      0.002857       0.0   
std        3.274488      1.75987      1.894498      0.414264       0.0   
min        0.000000      0.00000      0.000000      0.000000       0.0   
25%        0.000000      0.00000      0.000000      0.000000       0.0   
50%        0.000000      0.00000      0.000000      0.000000       0.0   
75%        0.000000      0.00000      0.000000      0.000000       0.0   
max      253.000000    253.00000    254.000000     62.000000       0.0   

       pixel781  pixel782  pixel783  
count   42000.0   42000.0   42000.0  
mean        0.0       0.0       0.0  
std         0.0       0.0       0.0  
min         0.0       0.0       0.0  
25%         0.0       0.0       0.0  
50%         0.0       0.0       0.0  
75%         0.0       0.0       0.0  
max         0.0       0.0       0.0  

[8 rows x 785 columns]

In [18]:
y_train = data['label']
X_train = data.iloc[:,1:]

print(y_train.shape)
print(X_train.shape)

(42000,)
(42000, 784)


### (1) Begin by fitting a random forest classifier using the full set of 784 explanatory variables and the model training set (train.csv). Record the time it takes to fit the model and then evaluate the model on the test.csv data by submitting to Kaggle.com. Provide your Kaggle.com score and user ID.

In [28]:
from sklearn.ensemble import RandomForestClassifier

In [29]:
start=datetime.datetime.now()

random_forest = RandomForestClassifier(n_estimators=10, max_features="sqrt", bootstrap=True, n_jobs=-1, random_state=42)
random_forest.fit(X_train, y_train)

end=datetime.datetime.now()
print(end-start)

0:00:02.442572


In [30]:
rf_predictions = random_forest.predict(test_data)
rf_predictions.reshape(-1,1)

array([[2],
       [0],
       [9],
       ...,
       [3],
       [9],
       [2]], dtype=int64)

In [36]:
print(rf_predictions.shape)
print(np.arange(1,28001).shape)

(28000,)
(28000,)


In [38]:
d = {'ImageId': np.arange(1,28001), 'Label': rf_predictions}
df = pd.DataFrame(data=d)

In [39]:
df.head()

ImageId  Label
0        1      2
1        2      0
2        3      9
3        4      7
4        5      3

In [47]:
df.to_csv('random_forest.csv', index=False)

In [48]:
print('Team: Brandi Beals')
print('Number: 2586')
print('Score: 0.93814')

Team: Brandi Beals
Number: 2586
Score: 0.93814


### (2) Execute principal components analysis (PCA) on the combined training and test set data together, generating principal components that represent 95 percent of the variability in the explanatory variables. The number of principal components in the solution should be substantially fewer than the 784 explanatory variables. Record the time it takes to identify the principal components.

In [50]:
print(X_train.head())
print(test_data.head())

   pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0   

In [51]:
full_data = pd.concat([X_train,test_data])

In [52]:
from sklearn.decomposition import PCA

In [54]:
start=datetime.datetime.now()

pca = PCA(n_components=0.95, random_state=42)
pca_model = pca.fit_transform(full_data)

end=datetime.datetime.now()
print(end-start)

0:00:14.779633


In [58]:
print(pca.n_components_)

154


### (3) Using the identified principal components from step (2), use the train.csv to build another random forest classifier. Record the time it takes to fit the model and to evaluate the model on the test.csv data by submitting to Kaggle.com. Provide your Kaggle.com score and user ID.

In [59]:
print(X_train.shape)
print(test_data.shape)
print(full_data.shape)

(42000, 784)
(28000, 784)
(70000, 784)


In [62]:
pca_X_train = pca_model[:42000]
print(pca_X_train.shape)

(42000, 154)


In [66]:
pca_X_test = pca_model[42000:]
print(pca_X_test.shape)

(28000, 154)


In [63]:
start=datetime.datetime.now()

random_forest_2 = RandomForestClassifier(n_estimators=10, max_features="sqrt", bootstrap=True, n_jobs=-1, random_state=42)
random_forest_2.fit(pca_X_train, y_train)

end=datetime.datetime.now()
print(end-start)

0:00:05.687586


In [67]:
rf_predictions_2 = random_forest_2.predict(pca_X_test)
print(rf_predictions_2.shape)
print(np.arange(1,28001).shape)

(28000,)
(28000,)


In [68]:
d2 = {'ImageId': np.arange(1,28001), 'Label': rf_predictions_2}
df2 = pd.DataFrame(data=d2)
df2.to_csv('random_forest_pca.csv', index=False)

In [69]:
print('Team: Brandi Beals')
print('Number: 2588')
print('Score: 0.88228')

Team: Brandi Beals
Number: 2588
Score: 0.88228


### (4) Submit both the RF Classifier and the PCA RF Classifier to Kaggle.com, and report both scores along with your user name.  I MUST have your user name to verify submission status.

My submissions can be found here: https://www.kaggle.com/brandibeals/competitions 

### (5) The experiment we have proposed has a MAJOR design flaw. Identify the flaw. Fix it. Rerun the experiment in a way that is consistent with a training-and-test regimen, and submit this to Kaggle.com. Provide your Kaggle.com score and user ID.

In [ ]:
X_train = pca.fit_transform(X_train)  
X_test = pca.transform(X_test) 